rust/storage/src/object_store.rs

```diff
@@ -0,0 +1,194 @@
+use std::sync::Arc;
+
+use chroma_error::ChromaError;
+use object_store::path::Path;
+use object_store::{GetOptions, ObjectStore as ObjectStoreTrait, PutOptions};
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio::sync::Mutex;
+
+use super::{GetError, PutError, StorageConfigError};
+
+#[derive(Clone)]
+pub struct ObjectStore {
+    object_store: Arc<dyn ObjectStoreTrait>,
+    upload_part_size_bytes: u64,
+    #[allow(dead_code)]
+    download_part_size_bytes: u64,
+}
+
+impl ObjectStore {
+    pub async fn try_from_config(
+        config: &super::config::ObjectStoreConfig,
+    ) -> Result<Self, Box<dyn ChromaError>> {
+        match &config.bucket.r#type {
+            super::config::ObjectStoreType::Minio => {
+                let object_store = object_store::aws::AmazonS3Builder::new()
+                    .with_region("us-east-1")
+                    .with_endpoint("http://minio.chroma:9000")
+                    .with_bucket_name(&config.bucket.name)
+                    .with_access_key_id("minio")
+                    .with_secret_access_key("minio123")
+                    .build()
+                    .map_err(|err| {
+                        tracing::error! {"Failed to create object store: {:?}", err};
+                        Box::new(StorageConfigError::InvalidStorageConfig) as _
+                    })?;
+                let object_store = object_store::limit::LimitStore::new(
```

Comment:
I think we will lose the request dedupe functionality of AdmissionControlledS3 if we switch to this

Reply:
I didn't look at its internals yet.  It was simply a nice-to-have I threw in at the last second.  I plan to make a wrapper out of our admission controller and internalize it to object store so that we get 1:1 with what we have today.  It also doesn't acquire parallel tokens.

Reply:
gotcha